Refactor UIColor Extensions

jrichardson0104 opened the pull request:
Updated tc_isEqualToColor to use objc method and updated
tc_colorWithHexString to throw an exception for invalid hex strings.
Updated test for tc_colorWithHexString for clarity and testing if
exception is thrown for invalid hex strings.

jrichardson0104 committed:
Refactored UIColor extensions

Updated tc_isEqualToColor to use objc method and updated
tc_colorWithHexString to throw an exception for invalid hex strings.
Updated test for tc_colorWithHexString for clarity and testing if
exception is thrown for invalid hex strings.

jrichardson0104 committed:
Merge branch 'develop' into refactor_UIColor_Extensions

nhgrif commented on ThunderCats/UIColor+TCAdditions.m:
```diff
@@ -28,12 +28,19 @@
+    {
+        [TCInvalidArgument raiseWithReason:@"invalid hex string"];
+    }
+    
```
We also need to make sure that the string only contains valid hexadecimal characters.  The digits 0-9, and the letters a-f (upper or lower case are both fine).

And arguably, even if we add this check in, we'd still be letting a string like `##0#123#4#5##` through because we just strip the `#` symbols out.  And I'd argue that this should return a valid color.

jrichardson0104 replied:
So, are you saying that we should raise an exception and not let '##0#123#4#5##' through?

nhgrif replied:
Correct.

nhgrif replied:
I'd recommend something like checking if the first character is `#`, and if so, strip off just that character.  After that, make sure the remaining characters are hex values.  Alternatively, a regex might work here.

nhgrif commented on ThunderCatsTests/UIColorTests.m:
```diff
@@ -30,101 +30,70 @@
+    [blackColor getRed:&red green:&green blue:&blue alpha:&alpha];
     
-    XCTAssertFalse(r1-r2 == 0 && g1-g2 == 0 && b1-b2 == 0 && a1-a2 == 0);
+    XCTAssertEqualWithAccuracy(red, 0.0/255.0, 0.05,  @"Expected red value of 0, but got %f", red);
```
Copy & pasted "red" all the way through here, I see.  Should be red, green, blue, and alpha.

nhgrif replied:
Accuracy could be even tighter.  Like `0.001`.

nhgrif commented on ThunderCatsTests/UIColorTests.m:
```diff
@@ -30,101 +30,70 @@
+    XCTAssertEqualWithAccuracy(red, 0.0/255.0, 0.05,  @"Expected red value of 0, but got %f", red);
+    XCTAssertEqualWithAccuracy(green, 0.0/255.0, 0.05,  @"Expected red value of 0, but got %f", green);
+    XCTAssertEqualWithAccuracy(blue, 0.0/255.0, 0.05,  @"Expected red value of 0, but got %f", blue);
+    XCTAssertEqual(alpha, 1,  @"Expected red value of 1, but got %f", alpha);
```
Alpha is also a `CGFloat` value and should also be asserted using `WithAccuracy`.

nhgrif commented on ThunderCatsTests/UIColorTests.m:
```diff
@@ -30,101 +30,70 @@
+    XCTAssertThrows([UIColor tc_colorWithHexString:@"00"]);
+    XCTAssertThrows([UIColor tc_colorWithHexString:@"F000"]);
+    XCTAssertThrows([UIColor tc_colorWithHexString:@"C123456"]);
+    XCTAssertThrows([UIColor tc_colorWithHexString:@"C123456789"]);
```
None of these tests try lower case hex characters (which should be valid).

nhgrif replied:
None of these tests try non hex characters which should be invalid.

For example, this test would fail:

```
XCTAssertThrows([UIColor tc_colorWithHexString:@"orange"]);
```

But it should pass.  Besides `a` and `e` those aren't valid hex characters.

nhgrif commented on ThunderCats/UIColor+TCAdditions.m:
```diff
@@ -28,12 +28,19 @@
+    
+    if ([cleanString length] != 3 && [cleanString length] != 6)
+    {
+        [TCInvalidArgument raiseWithReason:@"invalid hex string"];
```
This should be more like `[TCInvalidArgument raiseWithReason:@"%@ is not a valid hex color string.", hexString];`

nhgrif commented:
This pull request addresses [Issue #14](https://github.com/metova/ThunderCats/issues/14)

jrichardson0104 committed:
Refactored based on PR comments

Added lower case hex string test cases and cleaned up tests. Also,
added more checks to ensure a valid hex string before returning a
UIColor.

nhgrif commented on ThunderCats/UIColor+TCAdditions.m:
```diff
@@ -28,27 +28,42 @@
-    if ([cleanString length] == 6)
+    NSCharacterSet *validCharacters = [NSCharacterSet characterSetWithCharactersInString:@"0123456789ABCDEFabcdefg"];
+    
+    bool invalidLength = [hexString length] != 3 && [hexString length] != 6;
```
4 & 8 should also count as valid lengths.  Hex strings with a length of 4 or 8 are also specifying the alpha value of the color.

jrichardson0104 committed:
Added addition UIColor Method

Added a method to check if hex string is valid. Refactored
colorWithHexString to use this method and account for alpha values in
hex string.

jrichardson0104 committed:
Added more tests and refactored colorWithHexString

Tests for rgba and rrggbbaa formatted hex strings and a test for
isValidHexString.
